apps/query-ui/queryui/util.py

```diff
@@ -85,86 +143,136 @@ def docset_to_string(docset: DocSet) -> str:
-        data_list = []
-        directory = os.path.join(trace_dir, node_id)
+class QueryNodeTrace:
+    """Display query traces for a single query node."""
+
+    # The order here is chosen to ensure that the most important columns are shown first.
+    # The logic below ensure that additional columns are also shown, and that empty columns
+    # are omitted.
+    COLUMNS = [
+        "properties.path",
+        "properties.entity.accidentNumber",
+        "properties.key",
+        "properties.count",
+        "properties.entity.location",
+        "properties.entity.dateAndTime",
+        "properties.entity.aircraft",
+        "properties.entity.registration",
+        "properties.entity.conditions",
+        "properties.entity.windSpeed",
+        "properties.entity.visibility",
+        "properties.entity.lowestCeiling",
+        "properties.entity.injuries",
+        "properties.entity.aircraftDamage",
+        "text_representation",
+        "doc_id",
+        "parent_id",
+    ]
+
+    def __init__(self, trace_dir: str, node_id: str):
+        self.trace_dir = trace_dir
+        self.node_id = node_id
+        self.df = None
+        self.readdata()
+
+    def readfile(self, f):
+        with open(f, "rb") as file:
+            try:
+                doc = pickle.load(file)
+            except EOFError:
+                return None
+
+            # For now, skip over MetadataDocuments.
+            if "metadata" in doc.keys():
+                return None
+
+            # Flatten properties.
+            if "properties" in doc:
+                for prop in doc["properties"]:
+                    if isinstance(doc["properties"][prop], dict):
+                        for nested_property in doc["properties"][prop]:
+                            doc[".".join(["properties", prop, nested_property])] = doc["properties"][prop][
+                                nested_property
+                            ]
+                    else:
+                        doc[".".join(["properties", prop])] = doc["properties"][prop]
+                doc.pop("properties")
+
+            # Keep only the columns we care about.
+            # Limit size of each to avoid blowing out memory.
+            def format_value(value):
+                MAX_CHARS = 1024
+                if isinstance(value, str):
+                    if len(value) > MAX_CHARS:
+                        return value[:MAX_CHARS] + "..."
+                    else:
+                        return value
+                if isinstance(value, dict):
+                    return format_value(str(value))
+                if isinstance(value, list):
+                    return format_value(str(value))
+                return value
+
+            row = {k: format_value(doc.get(k)) for k in doc.keys() if k in self.COLUMNS or k.startswith("properties.")}
+            return row
+
+    def readdata(self):
+        directory = os.path.join(self.trace_dir, self.node_id)
```

Comment:
I like using the pathlib library for stuff like this:
```
directory = Path(self.trace_dir) / self.node_id
docs = [d for d in [self.readfile(f) for f in directory.glob("*") if f.is_file()] if d is not None]
```
does the same as the entire for filename loop